hal/common/retarget.cpp

```diff
@@ -94,24 +95,30 @@ FileHandle::~FileHandle() {
 }
 
 #if DEVICE_SERIAL
-extern int stdio_uart_inited;
-extern serial_t stdio_uart;
 #if MBED_CONF_CORE_STDIO_CONVERT_NEWLINES
 static char stdio_in_prev;
 static char stdio_out_prev;
 #endif
-#endif
 
-static void init_serial() {
-#if DEVICE_SERIAL
-    if (stdio_uart_inited) return;
-    serial_init(&stdio_uart, STDIO_UART_TX, STDIO_UART_RX);
+namespace mbed {
+
+Serial& get_stdio_serial()
+{
+    static bool stdio_uart_inited = false;
+    static Serial stdio_serial(STDIO_UART_TX, STDIO_UART_RX);
```

Comment:
Unfortunately, C++ is very terrible at garbage-collecting globally-allocated non-pod objects. It may not be much of a concern here, since we expect almost all devices to have stdio, but if a device never calls a serial function, this `static Serial` declaration will cause the entire serial codebase to get dragged into the final image.

This is what `SingletonPtr` was trying to solve, but it unfortunately doesn't support non-default constructors. 

If we want to avoid the class leak, we can use a static buffer with placement new. This also avoid multiple initialization flags introduced by the static constructor. It's ugly, but significantly reduces poorly-defined behaviour from my experience:

``` cpp
static bool stdio_inited = false;
static unsigned stdio_uart[(sizeof(Serial)+sizeof(unsigned)-1)/sizeof(unsigned)];

if (!stdio_inited) {
    new (stdio_uart) Serial(STDIO_UART_TX, STDIO_UART_RX);
#if MBED_CONF_CORE_STDIO_BAUD_RATE
    reinterpret_cast<Serial*>(stdio_uart)->baud(MBED_CONF_CORE_STDIO_BAUD_RATE);
#endif
    stdio_inited = true;
}

return reinterpret_cast<Serial*>(stdio_uart);
```